### ISSUE TITLE
Project architecture

### ISSUE THREAD

**Issue by se16n** [microsoft/Power-Fx#8]
Hello,

I'm very existed about this project! Thank you!

I'm wondering if this the only language spec is open sourced or an entire platform? To re-phrase it will be possible to use Power-Fx in self-hosted apps which doesn't use Power Platform at all?

**Comment by mblennegard** [comment]
@se16n 
It is only the Power Fx formula language that is being open-sourced.

Regarding self-hosted apps etc., I believe Microsoft has explicitly stated that they would like to see others implementing the Power Fx language in their own applications and tools. Again, this is only the Power Fx formula language.

**Comment by MikeStall** [comment]
@mblennegard  - it's our plan to eventually open source the PowerFx language _implementation_ too, and allow others to consume that implementation. Stay tuned.

**Comment by mi-hol** [comment]
So this issue is linked to #13

**Comment by ScatteredRay** [comment]
Can you provide any information about the PowerFX language implementation? While this is interesting to me, I have no idea if the implementation can be embedded in C++ applications, requires dotNet, or is web/JS based. Additionally, does it run interpreted or does it have a bytecode, is that bytecode standardized? What are the potentials for platform support.

**Comment by tomzhang** [comment]
totally agree with @ScatteredRay , we just need to know the fundamental requirement about what kind of language using for create PowerFX and basic architecture about it.

**Comment by jgrisham** [comment]
> Can you provide any information about the PowerFX language implementation? While this is interesting to me, I have no idea if the implementation can be embedded in C++ applications, requires dotNet, or is web/JS based. Additionally, does it run interpreted or does it have a bytecode, is that bytecode standardized? What are the potentials for platform support.

You are free to implement PowerFX support in whatever language _(or using whatever tools)_ you would like, but you would need to create your own interpreter or complier.

This project simply _defines_ PowerFX, much like a book defining HTML doesn’t require the existence of a web browser, let alone an implementation of one. Those familiar with IETF Internet RFCs might think of this as [draft-language-powerfx], the draft of a _specification_ RFC.

_(Put another way: one wouldn’t expect a book on Latin to come with the DNA for an ancient Roman citizen, nor should such a Roman be required to make some productive use of Latin.)_

PowerFX is primarily a functional language _(think Haskell)_, not a procedural one. It’s also very high-level. I don’t believe functional languages can be reduced directly to bytecode, since assembly / machine code on traditional processors is fundamentally procedural. If that is correct, then it will need an interpreter, just like PostScript or Python.

Tl;dr: the Microsoft code _implementing_ PowerFX in their _Power Platform_ is still proprietary and not available to the public.

**Comment by MikeStall** [comment]
We've now added the implementation! Announcement here: https://powerapps.microsoft.com/fi-fi/blog/power-fx-open-source-now-available/
